### ISSUE TITLE
Connection closed / logstash: Client Timeout / filebeat: EOF

### ISSUE THREAD

**Issue by jautz** [logstash-plugins/logstash-input-beats#163]
## Summary

In my experiments with the very promising filebeat/logstash setup for remote logging I ran into an issue with connections being closed prematurely. That is, when producing log data in intervals greater than five seconds the connection between filebeat and logstash would be closed. I did not witness any data loss but at least data duplication (only once so far).

Still I wanted to increase the timeout to something like a minute. I was able to locate logstash's "Client Timeout" debug message in source file `src/main/java/org/logstash/beats/BeatsHandler.java`.

I found the [FileBeat EOF Error discussion](https://discuss.elastic.co/t/filebeat-eof-error/60882) and the (slightly related) [logstash-plugins/logstash-input-beats@02bdcca commit](https://github.com/logstash-plugins/logstash-input-beats/commit/02bdccaf9bc6e86a2ba1517b5aaf6ab524c12d07) that mentions the `client_inactivity_timeout` setting that has been introduced with [Pull Request #108](https://github.com/logstash-plugins/logstash-input-beats/pull/108).

But the bottom line is: I failed to make it work. Either my increased timeout is simply ignored or I am looking in the wrong places. Any hints from your side will be highly appreciated.

So here is my setup and the reproducible steps that provoke error messages. Both machines (filebeat client and logstash server) run Debian Jessie.

## Server: logstash process accepting beats

```
$ apt-cache policy logstash
logstash:
  Installed: 1:5.0.1-1
  Candidate: 1:5.0.1-1
  Version table:
 *** 1:5.0.1-1 0
        500 https://artifacts.elastic.co/packages/5.x/apt/ stable/main amd64 Packages
        100 /var/lib/dpkg/status


$ cat /etc/logstash/pipeline.conf
input {
  beats {
    port => 5044
    # close idle connection after 1m (default: 15s)
    client_inactivity_timeout => 60
  }
}

output {
  file {
    path => "/var/log/logstash/timeout_demo.log"
  }
}


$ sudo /usr/share/logstash/bin/logstash --path.settings /etc/logstash -f /etc/logstash/pipeline.conf --log.level=debug
Sending Logstash's logs to /var/log/logstash which is now configured via log4j2.properties
```

## Client: filebeat process producing beats

```
$ apt-cache policy filebeat 
filebeat:
  Installed: 5.0.1
  Candidate: 5.0.1
  Version table:
 *** 5.0.1 0
        500 https://artifacts.elastic.co/packages/5.x/apt/ stable/main amd64 Packages
        100 /var/lib/dpkg/status


$ cat /etc/filebeat/filebeat.yml
filebeat.prospectors:
- input_type: log
  paths:
    - /tmp/shipme/*.log

filebeat.shutdown_timeout: 5
logging.metrics.enabled: false

output.logstash:
  hosts: ["vm-jj-lssrv.sandbox.lan:5044"]


$ sudo filebeat.sh -c /etc/filebeat/filebeat.yml -e
2016/11/23 17:20:42.799520 beat.go:264: INFO Home path: [/usr/share/filebeat] Config path: [/etc/filebeat] Data path: [/var/lib/filebeat] Logs path: [/var/log/filebeat]
2016/11/23 17:20:42.799907 beat.go:174: INFO Setup Beat: filebeat; Version: 5.0.1
2016/11/23 17:20:42.800245 logstash.go:90: INFO Max Retries set to: 3
2016/11/23 17:20:42.800536 outputs.go:106: INFO Activated logstash as output plugin.
2016/11/23 17:20:42.800842 publish.go:291: INFO Publisher name: vm-jj-lscl
2016/11/23 17:20:42.801371 async.go:63: INFO Flush Interval set to: 1s
2016/11/23 17:20:42.801785 async.go:64: INFO Max Bulk Size set to: 2048
2016/11/23 17:20:42.802120 beat.go:204: INFO filebeat start running.
2016/11/23 17:20:42.802152 registrar.go:66: INFO Registry file set to: /var/lib/filebeat/registry
2016/11/23 17:20:42.802202 registrar.go:88: INFO No registry file found under: /var/lib/filebeat/registry. Creating a new registry file.
2016/11/23 17:20:42.802222 crawler.go:34: INFO Loading Prospectors: 1
2016/11/23 17:20:42.802277 prospector_log.go:40: INFO Load previous states from registry into memory
2016/11/23 17:20:42.802292 prospector_log.go:67: INFO Previous states loaded: 0
2016/11/23 17:20:42.802371 crawler.go:46: INFO Loading Prospectors completed. Number of prospectors: 1
2016/11/23 17:20:42.802381 crawler.go:61: INFO All prospectors are initialised and running with 0 states to persist
2016/11/23 17:20:42.802414 prospector.go:106: INFO Starting prospector of type: log
2016/11/23 17:20:42.801719 logp.go:213: INFO Metrics logging disabled
2016/11/23 17:20:42.803244 registrar.go:211: INFO Starting Registrar
2016/11/23 17:20:42.803277 sync.go:41: INFO Start sending events to output
2016/11/23 17:20:42.803321 spooler.go:64: INFO Starting spooler: spool_size: 2048; idle_timeout: 5s
```

## Client: produce data in regular intervals

```
$ echo 5 > /tmp/shipme/.delay

$ ( while : ; do delay=$(</tmp/shipme/.delay); test -n "$delay" || exit 9; echo "$(date '+%F %T') producing data with $delay secs delay"; sleep $delay; done ) >> /tmp/shipme/data.log
```

## Server: watch incoming data

Ensure the setup works:
```
$ tail -n1 /var/log/logstash/timeout_demo.log
{"@timestamp":"2016-11-23T17:31:02.831Z","offset":954,"@version":"1","beat":{"hostname":"vm-jj-lscl","name":"v
m-jj-lscl","version":"5.0.1"},"input_type":"log","host":"vm-jj-lscl","source":"/tmp/shipme/data.log","message":"2016-11-23 18:31:02 producing data with 5 secs delay","type":"log","tags":["beats_input_codec_plain_applied"]}
```

Now start watching the logstash process's logfile for client timeout messages:
```
$ tail -F /var/log/logstash/logstash-plain.log | grep -i 'client timeout'
```

## Provoking the client timeout

So far the setup works like a charm. It ships log data for many happy hours without complaints.
Now we increase the delay to half a minute and see what happens.

```
(client)$ echo 30 > /tmp/shipme/.delay

tail on logstash host reports:
[2016-11-23T18:51:22,870][DEBUG][org.logstash.beats.BeatsHandler] Client Timeout
[2016-11-23T18:51:58,889][DEBUG][org.logstash.beats.BeatsHandler] Client Timeout
[2016-11-23T18:52:23,884][DEBUG][org.logstash.beats.BeatsHandler] Client Timeout
[2016-11-23T18:52:58,888][DEBUG][org.logstash.beats.BeatsHandler] Client Timeout
[...]

filebeat client reports:
2016/11/23 17:51:42.870608 sync.go:85: ERR Failed to publish events caused by: EOF
2016/11/23 17:51:42.871258 single.go:91: INFO Error publishing events (retrying): EOF
2016/11/23 17:52:07.872028 sync.go:85: ERR Failed to publish events caused by: EOF
2016/11/23 17:52:07.872074 single.go:91: INFO Error publishing events (retrying): EOF
2016/11/23 17:52:42.872883 sync.go:85: ERR Failed to publish events caused by: EOF
2016/11/23 17:52:42.872965 single.go:91: INFO Error publishing events (retrying): EOF
[...]
```

Switching back to a smaller delay (less than or equal to five seconds) and all is fine again.

**Comment by ph** [comment]
@jautz Thanks for the reproducable case, I've seen similar report too on this problem. I will do some testing and add an integration test for it.

**Comment by ph** [comment]
I believe this is a bug, the idleStateHandler should use the configured value and only be trigger every minute. 

OK, I have found the problem, the `client_inactivity_timeout` is not sent to the server.

https://github.com/logstash-plugins/logstash-input-beats/blob/master/lib/logstash/inputs/beats.rb#L172

Concerning duplicates, We are implementing at least once semantic and in this case FB is resending  the whole transmission when it detects something bad happened on the server side.

**Comment by ph** [comment]
I've create #164 that should fix the issue you are having, I will try to get that merged/released quickly.

**Comment by jautz** [comment]
Thank you, @ph, for your quick reaction. I am glad to hear you already found out what is causing this issue.

Concerning duplicates: I have read about the _at least once_ guarantee and wanted to confirm that this promise has been kept so far. Maybe I should have put it differently to make it sound more positive.

**Comment by ph** [comment]
@jautz I should have said *only once sementic* is really hard :), 

But there is way around that you can use the fingerprint or hashid filter to create a unique ID with the log data and use it as the document_id, this will make the es output update or create the event which should reduce the duplicates.

I've just pushed beats input 3.1.10 you can install it with bin/logstash-plugin update logstash-input-beats

**Comment by jautz** [comment]
@ph wrote:
> I've just pushed beats input 3.1.10 you can install it with bin/logstash-plugin update logstash-input-beats

Of course, I was eager to test the new version. However, the plugin update fails with the following message. I apologize if this is me doing things wrong since I am rather unexperienced with the elastic stack. As mentioned in the initial post this is a logstash 1:5.0.1-1 installed via apt from the elastic.co repositories.

```
# /usr/share/logstash/bin/logstash-plugin update logstash-input-beats                       
Updating logstash-input-beats
Error Bundler::InstallError, retrying 1/10
An error occurred while installing logstash-core-event-java (5.0.1), and Bundler cannot continue.
Make sure that `gem install logstash-core-event-java -v '5.0.1'` succeeds before bundling.
WARNING: SSLSocket#session= is not supported
Error Bundler::InstallError, retrying 2/10
An error occurred while installing logstash-core-event-java (5.0.1), and Bundler cannot continue.
Make sure that `gem install logstash-core-event-java -v '5.0.1'` succeeds before bundling.
WARNING: SSLSocket#session= is not supported
^C
```

If this is too much of a hassle, maybe I will wait for the next Debian package version that will contain this bugfix -- any idea when this would be?

**Comment by ph** [comment]
@jautz

**Comment by ph** [comment]
@jautz This could be a legit bug, let me try on my hand, 

Can you create a new issue for it and add the output of the following command?

```
DEBUG=1 JARS_DEBUG=true JARS_VERBOSE=true /usr/share/logstash/bin/logstash-plugin update logstash-input-beats  
```

**Comment by ph** [comment]
On a fresh debian install.

```
root@jessie:/usr/share/logstash# bin/logstash-plugin update logstash-input-beats
Updating logstash-input-beats
Updated logstash-input-beats 3.1.8 to 3.1.10
root@jessie:/usr/share/logstash#
```

**Comment by jautz** [comment]
Thank you for trying to reproduce. I can confirm that this must have something to do with my test server. I just tried it with a fresh docker container on my laptop using the logstash image from docker hub and there the update works, both with the builtin openjdk and also after replacing that with an oracle jdk.

With your debugging hints I will try to find out what's going wrong tomorrow. If it turns out to be some non-local problem I will open another issue for it.

**Comment by ph** [comment]
@jautz thanks its really appreciated! 👍

**Comment by makarchuk** [comment]
After updates seems to be an issue again. Downgraded plugin from 3.1.12 to 3.1.10 and it solved my problem

**Comment by edhill** [comment]
Just to add another data point - I agree with makarchuk that there seems to be a reversion.  I upgraded to a fresh 5.3.0 install and started seeing this problem again, and I downgraded the logstash-input-beats plugin to 3.1.10 and it fixed the problem for me.

**Comment by kholia** [comment]
I am running into this issue as well.

My Filebeat and Auditbeat logs contain lines like,

```
2018-06-30T01:21:11.061+0530    ERROR    logstash/async.go:235    Failed to publish events caused by: unexpected EOF
2018-06-30T01:21:11.062+0530    ERROR    logstash/async.go:235    Failed to publish events caused by: client is not connected
2018-06-30T01:21:12.064+0530    ERROR    pipeline/output.go:92    Failed to publish events: client is not connected
```

The output of `filebeat test output` and `auditbeat test output` commands is fine.
